### PR TITLE
[clang] Split up `SemaDeclAttr.cpp`

### PR DESCRIPTION
**Pull request by Endilll**
This patch moves language- and target-specific functions out of `SemaDeclAttr.cpp`. As a consequence, `SemaAVR`, `SemaM68k`, `SemaMSP430`, `SemaOpenCL`, `SemaSwift` were created (but they are not the only languages and targets affected).

Notable things are that `Sema.h` actually grew a bit, because of templated helpers that rely on `Sema` that I had to make available from outside of `SemaDeclAttr.cpp`. I also had to left CUDA-related in `SemaDeclAttr.cpp`, because it looks like HIP is building up on top of CUDA attributes.

This is a follow-up to #93179 and continuation of efforts to split `Sema` up. Additional context can be found in #84184 and #92682.

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -3825,6 +3843,115 @@ class Sema final : public SemaBase {
 
   void redelayDiagnostics(sema::DelayedDiagnosticPool &pool);
 
+  /// Diagnose mutually exclusive attributes when present on a given
```
Comment (erichkeane) [inline clang/include/clang/Sema/Sema.h:3846]: Where did this section of stuff come from?  At least the first 2 are attribute related and might be better in the Attr.td with a SemaRef, perhaps a few others as well.

File: clang/include/clang/Sema/SemaOpenMP.h [hunk 2]
```diff
@@ -32,6 +32,7 @@
 #include "clang/Basic/Specifiers.h"
 #include "clang/Sema/DeclSpec.h"
 #include "clang/Sema/Ownership.h"
+#include "clang/Sema/ParsedAttr.h"
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaOpenMP.h:35]: Could this be a forward decl?

File: clang/include/clang/Sema/SemaRISCV.h [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #include "clang/AST/Type.h"
 #include "clang/Basic/SourceLocation.h"
 #include "clang/Basic/TargetInfo.h"
+#include "clang/Sema/ParsedAttr.h"
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaRISCV.h:21]: Same question here.

File: clang/include/clang/Sema/SemaSYCL.h [hunk 4]
```diff
@@ -14,9 +14,11 @@
 #define LLVM_CLANG_SEMA_SEMASYCL_H
 
 #include "clang/AST/Decl.h"
+#include "clang/AST/DeclBase.h"
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaSYCL.h:17]: Wonder same about declbase as well, probably a bit of work on a few of these includes to do that sort of analysis.

File: clang/include/clang/Sema/SemaSwift.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  /// \returns true if the name is a valid swift name for \p D, false otherwise.
+  bool DiagnoseName(Decl *D, StringRef Name, SourceLocation Loc,
+                    const ParsedAttr &AL, bool IsAsync);
+  void AddParameterABIAttr(Decl *D, const AttributeCommonInfo &CI,
+                           ParameterABI abi);
+};
+
+} // namespace clang
+
+#endif // LLVM_CLANG_SEMA_SEMASWIFT_H
```
Comment (erichkeane) [inline clang/include/clang/Sema/SemaSwift.h:57]: needs newline.

**Review by erichkeane — COMMENTED** [review]
I didn't take a close look at the individual .cpp files, I assume that is mostly just Copy  & paste.

I would like us to be a little more forward-decl aggressive in the headers however, and perhpas try to minimize what gets pushed up to Sema.h.

File: clang/include/clang/Sema/Attr.h [hunk 6]
```diff
@@ -32,5 +36,87 @@ inline bool isFunctionOrMethodOrBlockForAttrSubject(const Decl *D) {
   return isFuncOrMethodForAttrSubject(D) || llvm::isa<BlockDecl>(D);
 }
 
+/// Return true if the given decl has a declarator that should have
+/// been processed by Sema::GetTypeForDeclarator.
+inline bool hasDeclarator(const Decl *D) {
+  // In some sense, TypedefDecl really *ought* to be a DeclaratorDecl.
+  return isa<DeclaratorDecl>(D) || isa<BlockDecl>(D) ||
+         isa<TypedefNameDecl>(D) || isa<ObjCPropertyDecl>(D);
```
Comment (AaronBallman) [inline clang/include/clang/Sema/Attr.h:44]:
```suggestion
  return isa<DeclaratorDec, BlockDecl, TypedefNameDecl, ObjCPropertyDecl>(D);
```

File: clang/include/clang/Sema/Attr.h [hunk 7]
```diff
@@ -32,5 +36,87 @@ inline bool isFunctionOrMethodOrBlockForAttrSubject(const Decl *D) {
...
+}
+
+/// hasFunctionProto - Return true if the given decl has a argument
+/// information. This decl should have already passed
+/// isFuncOrMethodForAttrSubject or isFunctionOrMethodOrBlockForAttrSubject.
+inline bool hasFunctionProto(const Decl *D) {
+  if (const FunctionType *FnTy = D->getFunctionType())
+    return isa<FunctionProtoType>(FnTy);
+  return isa<ObjCMethodDecl>(D) || isa<BlockDecl>(D);
```
Comment (AaronBallman) [inline clang/include/clang/Sema/Attr.h:53]:
```suggestion
  return isa<ObjCMethodDecl, BlockDecl>(D);
```

**Review by AaronBallman — COMMENTED** [review]
Precommit CI failures seem relevant.

File: clang/include/clang/Sema/SemaOpenMP.h (hunk 2 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaOpenMP.h]: Yeah, I totally forgot about forward decls while preparing this PR. I fixed that for the things added in this PR, leaving the rest for later.

File: clang/include/clang/Sema/SemaSwift.h (hunk 5 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaSwift.h]: Fixed.

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (Endilll) [reply clang/include/clang/Sema/Sema.h]: Making functions accept reference to `SemaBase` is novel, but seem to work fine without negative consequences. I moved the whole block to `Sema/Attr.h`.

File: clang/include/clang/Sema/SemaSYCL.h (hunk 4 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaSYCL.h]: This one should be fixed as well now.

File: clang/include/clang/Sema/Attr.h (hunk 6 above)
Comment (Endilll) [reply clang/include/clang/Sema/Attr.h:49]: We agreed to leave this out of scope of this PR, even though it makes perfect sense and should be done.

File: clang/include/clang/Sema/Attr.h (hunk 7 above)
Comment (Endilll) [reply clang/include/clang/Sema/Attr.h:58]: We agreed to leave this out of scope of this PR, even though it makes perfect sense and should be done.

File: clang/include/clang/Sema/SemaSYCL.h (hunk 4 above)
Comment (bader) [reply clang/include/clang/Sema/SemaSYCL.h]: > This one should be fixed as well now.

Can we drop `#include "clang/AST/Decl.h"` as well?

Comment (bader) [reply clang/include/clang/Sema/SemaSYCL.h]: I'm okay with that.
It just looks strange to add forward declaration for `Decl` class when full declaration is included with `#include "clang/AST/Decl.h"`. Probably the cleaner solution would be to drop forward declaration of `Decl` and add it in the follow-up patch.
Anyway, this is a minor comment and should not block the merge.

**Comment by RKSimon** [comment]
Would it make sense to add a new header (SemaUtils.h? SemaTargetUtils.h?) to move most of the exposed templated helpers to instead of Sema.h?

**Comment by Endilll** [comment]
> Would it make sense to add a new header (SemaUtils.h? SemaTargetUtils.h?) to move most of the exposed templated helpers to instead of Sema.h?

That's what `Sema/Attr.h` does, and those helpers for attribute handling went there.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by pogo59** [comment]
I wonder if a lot of the new target-specific headers don't need to be in clang/include.  That subtree is for headers that declare the exported interface (exported to other libs/layers); if the target-specific headers are just there for splitting up the module, they can stay in clang/lib/Sema.

**Pull request merged**